source/Examples/SmartCache/README.md

```diff
@@ -0,0 +1,115 @@
+
+It provides
+
+1. Shared memory communication channels via the [`Mlos.Agent`](../../Mlos.Agent) and [`Mlos.NetCore`](../../Mlos.NetCore) libraries.
+
+2. A [`Mlos.Agent.GrpcServer`](../../Mlos.Agent.GrpcClient/) GRPC channel to allow driving the experimentation process from a Jupyter notebook.
+
+3. A GRPC client to connect to the (Python) [`mlos.Grpc.OptimizerMicroserviceServer`](../../Mlos.Python/mlos/Grpc/OptimizerMicroserviceServer.py) to store and track those experiments.
+
+TODO: Diagrams
+
+## Building
+
+To build and run the necessary components for this example
+
+1. [Build the Docker image](../../../documentation/01-Prerequisites.md#build-the-docker-image) using the [`Dockerfile`](../../../Dockerfile) at the root of the repository.
+
+2. [Run the Docker image](../../../documentation/02-Build.md#create-a-new-container-instance) you just built.
+
+3. Inside the container, [build the compiled software](../../../documentation/02-Build.md#cli-make) with `make`:
+
+    ```sh
+    make dotnet-build cmake-build
+    ```
+
+    > This will build everything using a default `CONFIGURATION=Release`.
+    >
+    > To just build `SmartCache` and `Mlos.Agent.Server`, execute the following: \
+    > `make -C source/Examples/SmartCache && make -C source/Mlos.Agent.Server`
+
+4. For a `Release` build, the relevant output will be at:
+
+    - `Mlos.Agent.Server`:
+
+        `out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll`
+
+    - `SmartCache`:
+
+        `out/cmake/Release/source/Examples/SmartCache/SmartCache`
+
+    - `SmartCache.SettingsRegistry`
+
+        `out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU/SmartCache.SettingsRegistry.dll`
+
+## Executing
+
+> Note: these commands are given relative to the root of the MLOS repo.
+>
+> To move there, you can execute the following within the repository:
+>
+> `cd $(git rev-parse --show-toplevel)`
+
+`SmartCache` can be invoked separate, or by the `Mlos.Agent.Server` itself.
+
+Once started, `SmartCache` will attempt to register its component specific set of shared memory messages with the `Mlos.Agent` in the `Mlos.Agent.Server` using some `Mlos.Core` component registration messages.  That message includes the name of the `SettingsRegistry` assembly (`.dll`) corresponding to that component's settings/messages.
+
+The `Mlos.Agent.Server` needs to be told where it can find those assemblies.  To do that we provide an `MLOS_SETTINGS_REGISTRY_PATH` environment variable.
+
+In this case we populate it with the path to the `SmartCache.SettingsRegistry.dll`:
+
+```sh
+export MLOS_SETTINGS_REGISTRY_PATH="out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU:$MLOS_SETTINGS_REGISTRY_PATH"
+```
+
+Next, we can the `Mlos.Server.Agent` using the `dotnet` command:
+
+```sh
+tools/bin/dotnet out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll
```

Comment:
Does this output anything? If so, should we give the expected output here?
If not, we should change it to output "waiting for connection" or something like that.

Reply:
Yeah, that's a good idea.  I'll add that too.